src/main/java/org/spongepowered/api/data/manipulator/generator/CustomDataProvider.java

```diff
@@ -0,0 +1,65 @@
+package org.spongepowered.api.data.manipulator.generator;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import org.spongepowered.api.Sponge;
+import org.spongepowered.api.data.DataSerializable;
+import org.spongepowered.api.data.key.Key;
+import org.spongepowered.api.data.manipulator.DataManipulator;
+import org.spongepowered.api.data.manipulator.ImmutableDataManipulator;
+import org.spongepowered.api.data.value.BaseValue;
+import org.spongepowered.api.data.value.BoundedValue;
+import org.spongepowered.api.util.ResettableBuilder;
+
+import java.util.Arrays;
+
+public class CustomDataProvider {
+
+    public static RawBuilder builder() {
+        return Sponge.getRegistry().createBuilder(RawBuilder.class);
+    }
+
+    public static <T extends DataManipulator<T, I>, I extends ImmutableDataManipulator<I, T>> TypeBuilder<T, I> builder(Class<T> manipulatorClass, Class<I> immutableClass) {
+        return Sponge.getDataManager().createCustomBuilder(manipulatorClass, immutableClass);
+    }
+
+
+    public interface RawBuilder extends ResettableBuilder<DataRegistration<?, ?>, RawBuilder> {
+
+        RawBuilder withString(Key<? extends BaseValue<String>> key, String defaultValue);
```

Comment:
I feel like with* is well suited in this context, since we're adding "fields" to a "class." Then again, convention trumps all, doesn't it?